Add docs for `anchors` key (#7726)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-properties/anchors.md

---
title: anchors
sidebar_label: "anchors"
id: anchors
---

With YAML anchors, you can reuse configuration blocks across your dbt project files. Note that the way anchors are defined and used is different between dbt versions.

> dbt Core v1.9 supports standalone anchor definitions at the top level of YAML files. For example:

Author: Should we include this comparison with 1.9 syntax? Reference: https://github.com/dbt-labs/jaffle-minis/tree/main/minis/7a768_anchors

> dbt Core v1.9 supports standalone anchor definitions at the top level of YAML files. For example:

Reviewer: End of critical support for v1.9 will come this December and it will be removed from the docs. For the sake of maintenance, it might be beneficial to put this section in version blocks for lastversion="1.9"

<Tabs>
<TabItem value="model" label="Models">

<File name='models/_models.yml'>

```yml
columns: &id_column # standalone anchor definition - not supported in 1.10 and Fusion
  - name: id
    description: This is a unique identifier.
    data_tests:
      - not_null

models:
  - name: my_first_model
    columns: *id_column
```

</File>

</TabItem>

<TabItem value="source" label="Sources">

<File name='models/_sources.yml'>

```yml
sources:
  - &common_database # standalone anchor definition - not supported in 1.10 and Fusion
    name: common_database
    database: "{{ target.database }}"

  - <<: *common_database
    name: my_source
    schema: "{{ target.schema }}"
    tables:
      - name: my_seed
```

</File>

</TabItem>

</Tabs>

> In dbt Core v1.10 and dbt Fusion engine, standalone anchors that are meant to be used as reusable snippets should be moved under the `anchors:` key. For example:

Reviewer: Similar to the feedback above, it might be beneficial to put this section in version blocks with firstversion="1.10"

<Tabs>
<TabItem value="model" label="Models">

<File name='models/_models.yml'>

```yml
anchors: # use the anchors key
  - columns: &id_column
    - name: id
      description: This is a unique identifier.
      data_tests:
        - not_null

models:
  - name: my_first_model
    columns: *id_column
  - name: my_second_model
    columns: *id_column
```

</File>

</TabItem>

<TabItem value="source" label="Sources">

<File name='models/_sources.yml'>

```yml
anchors: # use the anchors key
  - &common_database
    name: common_database
    database: "{{ target.database }}"

sources:
  - <<: *common_database
    name: my_source
    schema: "{{ target.schema }}"
    tables:
      - name: my_seed
```

</File>

</TabItem>

</Tabs>

:::important 

> Not all anchors should be moved under an `anchors:` block. Some anchors are part of the main YAML structure (for example, defining tests on a column) and should _not_ be moved under a top-level `anchors:` key. For example:

Author: Please review accuracy of this callout.

> Not all anchors should be moved under an `anchors:` block. Some anchors are part of the main YAML structure (for example, defining tests on a column) and should _not_ be moved under a top-level `anchors:` key. For example:

Reviewer: Same feedback as the upgrade guide
```suggestion
Not all anchors should be moved under an `anchors:` block, such as anchors that are part of the main YAML structure (for example, defining tests on a column).

For example:
```

---
title: anchors
sidebar_label: "anchors"
id: anchors
---

> With YAML anchors, you can reuse configuration blocks across your dbt project files. Note that the way anchors are defined and used is different between dbt versions.

Reviewer 2: ```suggestion
Anchors are a [feature of YAML](https://yaml.org/spec/1.2.2/#692-node-anchors) making it possible to reuse configuration blocks inside of a single YAML file. Fragments of configuration which are not valid on their own or which only exist as template data should be enclosed in the `anchors:` key, to protect them from being rejected during file validation:

- The dbt Fusion engine will raise an error when it encounters unexpected keys, including those used as part of an anchor definition. _Note that during the Fusion engine's beta period, these errors are downgraded to warnings._
- Version 1.10 and later of dbt Core will raise a warning on unexpected keys.
- Earlier versions of dbt Core will not raise a warning at all
```

- That link to the YAML spec is confusing as hell but there's not really anything else good to link to, and I feel like it's useful to make clear that they're part of the YAML spec, not our own invention
- I can’t make 1.10 raise a warning on unexpected keys, there might be a behaviour change flag I haven't set but I would want to have someone validate this.

<VersionBlock lastVersion="1.9">

> dbt Core v1.9 supports standalone anchor definitions at the top level of YAML files. For example:

Reviewer 2: This is not really the right framing - it's actually that Fusion will complain when it sees unexpected keys, and sometimes those unexpected keys are standalone anchors.

Author: Revised the wording here: https://github.com/dbt-labs/docs.getdbt.com/pull/7726/commits/b1274db2a928085f4f5bf8c1ff6a4035d075d8da#diff-bf07beb7a4e1d2583641034f4fc526a9799c02a25ad765b6639cceca0c52d8b8

The following example creates an anchor whose alias is `*id_column_alias`. The `id` column, its description, data type, and data tests are all applied to `my_first_model`, `my_second_model`, and `my_third_model`.

<File name='models/_models.yml'>

```yml
anchors: 
  - &id_column_alias
      name: id
      description: This is a unique identifier.
      data_type: int
      data_tests:
        - not_null
        - unique

models:
  - name: my_first_model
    columns: 
      - *id_column_alias
      - name: unrelated_column_a
        description: This column is not repeated in other models.
      - name: unrelated_column_b
  - name: my_second_model
    columns: 
      - *id_column_alias
      - name: unrelated_column_c
  - name: my_third_model
    columns: 
      - *id_column_alias
      - name: unrelated_column_d
```

</File>

<Lightbox src="/img/reference/resource-properties/anchor_example_expansion.png" title="Behind the scenes, the alias is replaced with the object defined by the anchor."/>

### Merge syntax

> Sometimes, an anchor is mostly the same but one part needs to be overridden. When the anchor refers to a dictionary/mapping (not a list or a [scalar value](https://softwareengineering.stackexchange.com/questions/238033/what-does-it-mean-when-data-is-scalar/238045#238045)), you can use the `<<:` merge syntax to override an already-defined key, or add extra keys to the dictionary. For example:

Author: Instead of adding a link here, maybe we can add a hoverSnippet to define this term? Does this definition look accurate: "A single piece of data (for example, a number, character, string, or object) rather than a collection or set of values."

Reviewer 2: Yes that would be great - I thought we used to have a glossary but I couldn't find it. I'd say `A single piece of data (for example, a number or string) rather than a collection or set of values.`

        config:
          meta: 
            extra_key: extra_value # adds config.meta.extra_key to just this version of the id column, in addition to the name, description, data type, and data tests
      - name: unrelated_column_d

sources:
  # both sources start with their database, loader, and freshness expectations set from the anchor, and merge in additional keys
  - <<: *source_template_alias
    name: salesforce
    schema: etl_salesforce_schema
    tables:
      - name: opportunities
      - name: users
  - <<: *source_template_alias
    name: hubspot
    schema: etl_hubspot_schema
    tables:
      - name: contacts
```

</File>

## Tips and common pitfalls

- You can't merge additional elements into a list which was defined as an anchor. For example, if you define an anchor containing multiple columns, you can't attach extra columns to the end of the list. Instead, define each column as an individual anchor and add each one to the relevant tables.
<VersionBlock lastVersion="1.9">
- Old versions of dbt Core (v1.9 and earlier) do not have a dedicated `anchors:` key. If you need to define a standalone anchor, you can leave it at the top level of your file.
</VersionBlock>

> - You do not need to move existing anchors to the `anchors:` key if they are already defined in a larger valid YAML object. dbt Core v1.10 and higher will raise a warning if it finds invalid anchors, whereas the dbt Fusion engine will raise an error once it leaves beta. For example, the following `&customer_id_tests` anchor does not need to be moved, because it is also a valid part of the existing `columns` block.

Author: Since Fusion is still in beta, should we also mention here that it will raise a warning?

Reviewer 2: I handwaved over this because technically Fusion is higher than v1.10 and I didn't think it mattered that much, especially since we hope to move to raising errors very soon.

---
title: anchors
sidebar_label: "anchors"
id: anchors
---

> Anchors are a [YAML feature](https://yaml.org/spec/1.2.2/#692-node-anchors) that make it possible to reuse configuration blocks inside a single YAML file. Fragments of configuration which are not valid on their own or which only exist as template data should be enclosed in the `anchors:` key, which was introduced in dbt Core v1.10. Using the `anchors:` key protects these fragments from being rejected during file validation.

Reviewer 3: ```suggestion
Anchors are a [YAML feature](https://yaml.org/spec/1.2.2/#692-node-anchors) that let you reuse configuration blocks inside a single YAML file. In dbt Core v1.10, the `anchors:` key was introduced to enclose configuration fragments that aren't valid on their own or that only exist as template data. Using the `anchors:` key ensures these fragments won't be rejected during file validation.
```

> Anchors are a [YAML feature](https://yaml.org/spec/1.2.2/#692-node-anchors) that make it possible to reuse configuration blocks inside a single YAML file. Fragments of configuration which are not valid on their own or which only exist as template data should be enclosed in the `anchors:` key, which was introduced in dbt Core v1.10. Using the `anchors:` key protects these fragments from being rejected during file validation.

Reviewer 3: SHould we add "Description" title to math other reference pages?

In dbt Core v1.10 and higher, invalid anchors trigger a warning. In the dbt Fusion engine, these invalid anchors will result in errors when Fusion leaves beta.

<VersionBlock lastVersion="1.9">
:::note
You can define anchors in dbt Core v1.9 and earlier, but there is no dedicated location for anchors in these versions. If you need to define a standalone anchor, you can put it at the top level of your YAML file.
:::
</VersionBlock>

## YAML anchor syntax

### Anchors and aliases

> To define a YAML anchor, add an `anchors:` block in your YAML file and use the `&` symbol in front of the anchor's name (for example, `&id_column_alias`). This creates an alias which you can reference elsewhere by prefixing the alias with a `*` character.

Reviewer 3: Thanks for this explanation! It's super helpful!

models:
  - name: my_first_model
    columns: 
      - *id_column_alias # brings in the full anchor defined above
      - name: unrelated_column_a
        description: This column is not repeated in other models.
      - name: unrelated_column_b
  - name: my_second_model
    columns: 
      - <<: *id_column_alias
        data_type: bigint # overrides the data_type from int to bigint, while inheriting the name, description, and data tests
      - name: unrelated_column_c
  - name: my_third_model
    columns: 
      - <<: *id_column_alias
        config:
          meta: 
            extra_key: extra_value # adds config.meta.extra_key to just this version of the id column, in addition to the name, description, data type, and data tests
      - name: unrelated_column_d

sources:
  # both sources start with their database, loader, and freshness expectations set from the anchor, and merge in additional keys
  - <<: *source_template_alias
    name: salesforce
    schema: etl_salesforce_schema
    tables:
      - name: opportunities
      - name: users
  - <<: *source_template_alias
    name: hubspot
    schema: etl_hubspot_schema
    tables:
      - name: contacts
```

</File>

> ## Tips and common pitfalls

Reviewer 3: This is great but not a common title/section I see in other reference pages. I did see "Usage" and "Usage notes" in some sections. Another option is rewriting these as FAQs. 

I think it's important to be consistent (for humans and agents) and we have some work in References to do that.

<VersionBlock lastVersion="1.9">

> - Old versions of dbt Core (v1.9 and earlier) do not have a dedicated `anchors:` key. If you need to define a standalone anchor, you can leave it at the top level of your file.

Reviewer 3: I wonder if this should be left unversioned? Seems like important information we'd want everyone to see.
